commands: add `rbfpsbt` command

jp1ac4 opened the pull request:
This PR relates to #43 and #236.

It adds a `rbfpsbt` command that generates a PSBT to replace an existing transaction using RBF. This replacement can either preserve non-change outputs and simply bump fees or remove non-change outputs and effectively cancel the transaction. The inputs and change output may need to be updated in accordance with the higher fee.

I've also added a `getmempoolentry` call to the bitcoin interface that is used for checking information about descendant transactions.

To facilitate development, I've made some temporary changes in the GUI so that replacement PSBTs can be signed and broadcast, but these changes might not be part of this PR in the end.

jp1ac4 commented on src/bitcoin/d/mod.rs:
```diff
@@ -1394,3 +1411,85 @@ impl<'a> CachedTxGetter<'a> {
 }
+
+#[derive(Debug, Clone)]
+pub struct GetMempoolEntryRes {
```
I haven't included all the fields in the `getmempoolentry` response yet.

darosior replied:
No need to include the fields we won't be using (see how we only include those in every other command result).

jp1ac4 replied:
Do you think it's worth removing those I've already added that are not currently being used?

darosior replied:
Yes. Dead code is bad.

jp1ac4 commented on src/commands/mod.rs:
```diff
@@ -736,6 +774,370 @@ impl DaemonControl {
     }
 
+    // This is like `create_spend` with slightly different parameters.
+    fn create_rbf_spend(
```
For now, this is almost the same as `create_spend` with some slightly different parameters. Perhaps a follow-up PR could refactor to reduce duplicate code.

jp1ac4 replied:
After https://github.com/wizardsardine/liana/pull/821 with the addition of a change parameter, the only real difference between this method and `create_spend` will be:
- a `min_fee` parameter (I think we can avoid this parameter by just calling this function with an increasing feerate until it has the requried fee, which I'm already doing anyway given that the min fee depends on the size of the replacement transaction itself, see https://github.com/wizardsardine/liana/pull/816/files#diff-bf2f650cdc66abd13076c47b25e8c9f783f06ec335bbd30822fdd690547615fdR1100)
- whether to use a mix of mandatory and optional candidates or just those in `coins_outpoints` -> we could change the logic of `create_spend` so that optional candidates are used not when `coins_outpoints` is empty but when this extra Boolean parameter is `true`.

So I think by adding a Boolean parameter to `create_spend` in addition to #821, we could remove `create_rbf_spend`.

darosior replied:
> a min_fee parameter (I think we can avoid this parameter by just calling this function with an increasing feerate until it has the requried fee

I don't think we should do that.

But we can still have a `create_spend_internal` function which takes more parameter and is called by both `create_spend` and `rbf_psbt`.

darosior replied:
Looks good to me!

ICYMI: https://github.com/wizardsardine/liana/pull/816/files#diff-bf2f650cdc66abd13076c47b25e8c9f783f06ec335bbd30822fdd690547615fdR465-R466 the comment here is now outdated.

jp1ac4 replied:
Great, thanks for the quick feedback. I'll update that comment and check the others.

darosior commented on src/bitcoin/d/mod.rs:
```diff
@@ -1134,6 +1134,23 @@ impl BitcoinD {
     }
 
+    /// Get mempool entry.
+    // TODO: Should existing calls to `is_in_mempool` use this function instead and check is_some()?
```
Yes. :)

jp1ac4 replied:
That's done now. Do you think it should be called `get_mempool_entry()` or `mempool_entry()`, or perhaps a mixture of both depending on where the method is?

darosior replied:
I generally follow this rule: longer but explicit / clear is better than shorter but implicit / less clear. If clear already, shorter is better. In this case i think `mempool_entry()` is a better fit.

jp1ac4 commented on src/commands/mod.rs:
```diff
@@ -736,6 +774,372 @@ impl DaemonControl {
+        // sum of fees paid by original transactions + incremental feerate * replacement size.
+        for i in 0..100_u64 {
+            println!("trying feerate of {} with i {}", feerate_vb + i, i);
+            let rbf_psbt = self.create_rbf_spend(
```
In case `change_index` is `None`, I can pass an arbitrary value here to avoid incrementing the change index while we try different feerates, and then once the necessary feerate has been found, we can do a final call that increments the change index if necessary.

jp1ac4 replied:
Instead of this initial idea, I instead pass the next change address directly to `create_spend_internal`. It might be that we still increment the change index once in an intermediate call and then the final transaction won't have change, but this approach seems simpler.

darosior commented on src/commands/mod.rs:
```diff
@@ -736,6 +774,372 @@ impl DaemonControl {
+                must_select: true,
+            })
+            .collect();
+        if !destinations.is_empty() {
```
Why not including them if it's a self-spend?

jp1ac4 replied:
My thinking was that for a self-send, we only want to consider the same coins again, in the same way that we don't use coin selection for a new self-send. But perhaps for RBF, allowing to add an extra coin to cover the fee would make sense.

darosior replied:
Yes i think it should be expected that if you are going to RBF, some new inputs may be added to the transaction. We should warn the user in the documentation. And maybe also in the GUI but i'm not sure how not to confuse the user who doesn't care (most users don't i think).

jp1ac4 replied:
I added a comment to the API doc that additional coins may be used, including for self-sends.

darosior commented on src/commands/mod.rs:
```diff
@@ -736,6 +774,372 @@ impl DaemonControl {
+                    }),
+            );
+        }
+        // Try with increasing feerate until fee paid by replacement transaction is high enough.
+        // Replacement fee must be at least:
+        // sum of fees paid by original transactions + incremental feerate * replacement size.
+        for i in 0..100_u64 {
```
Oof. No. You should simply compute the minimum fee of the replacement instead of blindly trying out increasing feerates.

darosior reviewed (commented):
I'm a bit baffled by your iterating approach to get the minimum absolute fee.

jp1ac4 commented on src/commands/mod.rs:
```diff
@@ -736,6 +774,372 @@ impl DaemonControl {
+        let prev_psbt = db_conn
+            .spend_tx(txid)
+            .ok_or(CommandError::UnknownSpend(*txid))?;
+        let prev_coins = {
+            let prev_outpoints: &Vec<bitcoin::OutPoint> = &prev_psbt
+                .unsigned_tx
+                .input
+                .iter()
+                .map(|txin| txin.previous_output)
+                .collect();
+            let prev_coins = db_conn.coins_by_outpoints(prev_outpoints);
+            assert_eq!(prev_outpoints.len(), prev_coins.len());
+            prev_coins
+        };
```
Yep, I found it was easier this way to keep track of which variables are needed later on. But in case you'd prefer consistency with the rest of the code, I could also avoid doing that.

darosior replied:
Interesting. No, you can keep doing it.

jp1ac4 commented on src/commands/mod.rs:
```diff
@@ -736,6 +774,372 @@ impl DaemonControl {
+                self.config.bitcoin_config.network,
+            )
+            .map_err(CommandError::Address)?;
+            match db_conn.derivation_index_by_address(&address) {
+                Some((ind, true)) => {
+                    // TODO: check if there is more than one change output?
+                    println!("found change at address {} and index {}", address, ind);
+                    change_index = Some(ind);
+                }
```
My general assumption has been that change addresses are added automatically to the transaction as change output and not manually used as destinations by the user, but that's not guaranteed, even if a single change address is used. Perhaps in case there's more than one change address used, we could preserve all of them and not treat any as change output.

darosior replied:
> As discussed, if more than one of our change addresses is used in the existing transaction, we can keep one of them to use for change in the replacement transaction (and drop the others?).

No i don't think we should drop the others. We should just keep them like any other destinations.

> Perhaps the choice of change address to use in the replacement should not depend on its value, but rather its index? E.g. choosing the highest index so that there's less chance of reusing change addresses in a future transaction.

That's a good point but i think it's still better to choose the largest value change output, to take the fees from there to avoid having to increase the transaction size.

----

In general my thinking is: if the user has set the outputs the way they are, he has a reason. Let's avoid touching the structure of the outputs *except* if we can find a change output suitable to be reduced in order to pay for the fees, to avoid including more inputs and/or an additional change output.

jp1ac4 replied:
I changed the logic so that it chooses the change address with the highest value and then the one with the highest index, in case there's more than one. The other change addresses are then treated in the same way as non-change outputs.

jp1ac4 commented on src/commands/mod.rs:
```diff
@@ -736,6 +774,372 @@ impl DaemonControl {
+        }
+        // We take all previous inputs as mandatory candidates and, if not self-send, include
+        // confirmed coins as optional.
+        // TODO: For `is_cancel`, shall we include just one previous input?
```
OK yep that makes sense. I think starting with the largest would be good.

jp1ac4 replied:
I've changed the `is_cancel` case so that in the first instance, it uses the previous coins as optional candidates and we let the coin selection decide which of them are best to use given the min fee and feerate. If there are insufficient funds using just the previous coins as candidates, it then uses the same approach as for `!is_cancel`, i.e. all previous coins are mandatory candidates and confirmed coins are optional.

jp1ac4 replied:
This approach for `is_cancel` using optional candidates doesn't always work. The replacement transaction is a self-send, but it can happen if we have optional candidates that coin selection gives us a changeless solution without the optional candidates rather than one with change using the optional candidates. I think I'll need to use just mandatory candidates, adding one at a time as originally suggested until we have a solution with change.

jp1ac4 replied:
I think this could also affect the case where we want to bump the fee of a self-send: the higher fee may lead to a solution with no change. I'll use a similar approach for that case too.

One of the available scoring metrics for coin selection is [`Changeless`](https://github.com/evanlinjin/bdk/blob/d620dc61e809a8c280d623d5275abe8413b4b00a/nursery/coin_select/src/metrics/changeless.rs#L5). In principle, we might be able to use a similar `WithChange` metric by editing this [line](https://github.com/evanlinjin/bdk/blob/d620dc61e809a8c280d623d5275abe8413b4b00a/nursery/coin_select/src/metrics/changeless.rs#L18) to end with `is_some()` instead of `is_none()`. Then we could just pass the mandatory and optional candidates without having to increment the number of candidates.

jp1ac4 replied:
I did a quick test using a modified version of the `LowestFee` metric:
```
pub struct LowestFeeWithChange<'c, C> {
    pub lowest_fee: LowestFee<'c, C>, // This is the metric we're currently using
    pub must_have_change: bool,
}
```
 so that its score is given by:
```
if cs.is_target_met(self.lowest_fee.target, drain)
    && (drain.is_some() | !self.must_have_change)
{
    self.lowest_fee.score(cs) // We reuse the `LowestFee` score function.
} else {
    None
}
```
So `LowestFeeWithChange` is basically the same as `LowestFee`, except that `score()` will return `None` in case there's no change (depending on `must_have_change`).

From my tests, this avoids the problems I was facing with changeless solutions and also avoids the need for the extra logic adding candidates one at a time.

jp1ac4 replied:
Hmm I think it's a bit more complicated as we would also need to change how `bound()` is implemented. I'll continue with the iterating approach for now, adding candidates one by one.

jp1ac4 replied:
Sorry for all the noise, but I think this approach is still working. I just had to apply the same logic to the fallback in case BnB gave no solution. I've made the changes in a separate branch:
https://github.com/wizardsardine/liana/compare/master...jp1ac4:liana:lowest-fee-with-change-metric

It may be worth considering this approach as it should be simpler than iterating over the set of candidates as well as the min fee.

darosior replied:
This (the code and the explanation) makes sense to me.

It reminds me of the case where an output has more than one output paying to one of our addresses. Another approach could be to just set the output(s) as part of the destination, but it's ugly as it might create another change output while there is already one (or more). On the other hand it would conserve the value of the output(s). Can this ever be a reasonable expectation of the user? Hmm i don't think it'd be worth risking further duplication of the change.
TL;DR: i think you should go with your with-change metric.

jp1ac4 replied:
I also considered an approach of treating the change as a regular destination in order to avoid a changeless solution, but had similar concerns.

OK great, I'll add the with-change metric in a separate PR.

jp1ac4 replied:
PR for new metric: https://github.com/wizardsardine/liana/pull/832.

jp1ac4 commented on src/commands/mod.rs:
```diff
@@ -736,6 +774,372 @@ impl DaemonControl {
+            .ok_or(CommandError::RbfError(RbfErrorInfo::TxNotYetBroadcast(
+                *txid,
+            )))?;
+        // Check sum of descendant counts of all directly conflicting transactions is not too high.
+        // We are assuming we only have one directly conflicting transaction.
+        if mempool_entry.descendant_count > RBF_MAX_DESCENDANT_COUNT {
+            return Err(CommandError::RbfError(RbfErrorInfo::TooManyDescendants(
+                mempool_entry.descendant_count,
+            )));
+        }
```
Ah yes, I should also check the ancestor count. Perhaps also ancestor/descendant sizes.

darosior replied:
Another option is to not try to exhaustively match the current RBF policy of Bitcoin Core, and just let it fail in `sendrawtransaction` below? I don't think there is much benefit to erroring early, and being exhaustive would mean investing some time into it. Probably not worth it.

darosior replied:
Sounds good.

jp1ac4 commented:
I rebased on master and updated the functional test for the new `rbfpsbt` command, which depends on the fix from #829.

Other changes relating to the comments above are still ongoing.

darosior commented on tests/test_rpc.py:
```diff
@@ -16,6 +16,7 @@
     get_txid,
     spend_coins,
     sign_and_broadcast,
+    sign_and_broadcast_psbt,
```
Thanks for cleaning up my mess. :)

There is already a `sign_and_broadcast` as you can see so maybe we should avoid duplicating them. But not here. Let's not scope creep this PR.

darosior commented on gui/src/app/state/transactions.rs:
```diff
@@ -87,7 +87,11 @@ impl State for TransactionsPanel {
                 Ok(txs) => {
                     self.warning = None;
-                    for tx in txs {
+                    // FIXME: This is a quick hack to ensure order is consistent in order to quickly
+                    // check and switch between transactions.
+                    let mut txs_sorted = txs;
+                    txs_sorted.sort_by_key(|tx| tx.tx.txid());
+                    for tx in txs_sorted {
```
Seems to be worth on its own?

jp1ac4 replied:
There might be a better way to order the transactions, and possibly other changes that would be helpful when viewing the transactions list, so shall we leave this for a separate PR?

I think the only GUI change required for RBF, apart from the negative fees issue for missing coins that will be addressed in #820 and #825, is to leave the replacement PSBT status as "Pending" instead of "Deprecated" so that it can be signed and broadcast. I've made that change, but still need to check if we can set the replaced PSBT to "Deprecated" once the replacement has been confirmed.

darosior replied:
Sure.

jp1ac4 replied:
PSBTs with input coins that have been spent by another confirmed transaction will be shown as Deprecated.

I've made a further change so that PSBTs with only inputs that no longer exist are shown as Deprecated. This can happen if the PSBT was spending the outputs of a transaction that was replaced or otherwise no longer exists.

darosior commented on gui/src/daemon/model.rs:
```diff
@@ -129,7 +133,14 @@ impl SpendTx {
-            fee_amount: inputs_amount - spend_amount - change_amount,
+            // FIXME: this is a quick hack to deal with descendant PSBTs whose ancestor has been replaced.
+            // At least one input coin no longer exists so fee would be negative.
+            // Perhaps such PSBTs should be deleted or hidden?
```
I think we should keep them but display a "Can't compute fees" placeholder. xref https://github.com/wizardsardine/liana/issues/820 and https://github.com/wizardsardine/liana/pull/825.

darosior reviewed (commented):
Looking good!

jp1ac4 commented on src/commands/mod.rs:
```diff
@@ -736,6 +774,372 @@ impl DaemonControl {
+                .expect("fee must be Ok as transaction is already in mempool");
+            let prev_tx_size: u64 = prev_psbt
+                .clone()
+                .extract_tx()
```
I've updated to use the mempool entry for vsize and the fee. I'm using `base` from the result, but there's also `modified`. I'm not sure which of these is used by bitcoin core in the context of checking the directly conflicting transactions' feerates for RBF.

darosior replied:
`base` is the one we should use. :+1:

darosior commented on src/commands/mod.rs:
```diff
@@ -736,6 +753,403 @@ impl DaemonControl {
+        // Replacement fee must be at least:
+        // sum of fees paid by original transactions + incremental feerate * replacement size.
+        // Loop will continue until either we find a suitable replacement or we have insufficient funds.
+        let mut rbf_vsize = 0_u64;
+        loop {
+            // Note that `rbf_vsize` strictly increases on each iteration as otherwise `min_fee`
+            // will ensure that the PSBT meets the required replacement fee and the loop will exit.
+            let min_fee = mempool_entry.fees.descendant.to_sat() + rbf_vsize;
```
Why start at 0? This will always fail.

jp1ac4 replied:
I think with 0 we could still get a replacement with a high enough fee due to the higher feerate, and also in case the change output falls below the dust threshold and is added to fees.

darosior replied:
Fair enough.

jp1ac4 commented on src/commands/utils.rs:
```diff
@@ -244,3 +244,26 @@ pub fn select_coins_for_spend(
+///
+/// `max_sat_weight` is the maximum weight difference of an input in the
+/// transaction before and after satisfaction. Must be in weight units.
+pub fn unsigned_tx_max_vbytes(tx: &bitcoin::Transaction, max_sat_weight: u64) -> u64 {
```
This function could be used in `sanity_check`, but I'll wait in case any changes are required here (we could also change `sanity_check` in a separate PR).

jp1ac4 commented on src/commands/mod.rs:
```diff
@@ -736,6 +774,372 @@ impl DaemonControl {
+                .ok_or(CommandError::InsaneFees(InsaneFeeInfo::InvalidFeerate))?
+        };
+        // Check replacement transaction's feerate is greater than previous feerate (the only directly conflicting transaction).
+        // TODO: Should we throw error or just set feerate_vb = prev_feerate_vb + 1 and check `rbf_psbt` below?
```
I've added some logic along these lines, although it might need some tweaking.

jp1ac4 commented on src/commands/mod.rs:
```diff
@@ -780,6 +840,229 @@ impl DaemonControl {
+            .filter_map(|c| {
+                // In case the user attempts RBF before the previous coins have been updated in the DB,
+                // they would be returned as confirmed and not spending.
+                if !prev_coins.contains_key(&c.outpoint) {
```
An alternative would be to wait for the spend info of all coins to be updated.

jp1ac4 commented on src/commands/mod.rs:
```diff
@@ -780,6 +840,229 @@ impl DaemonControl {
+                // transaction, then set all previous coins as mandatory and add confirmed coins as
+                // optional, unless we have already done this.
+                Err(CommandError::CoinSelectionError(_))
+                    if is_cancel && candidate_coins.iter().all(|c| !c.must_select) =>
```
An alternative approach later on might be for coin selection to pick at least one from a subset. E.g. we could use a metric that only returns a score if the selection includes at least one candidate from this subset. So for the cancel case, we would want at least one from the previous inputs, and then zero or more from the optional.

darosior commented on src/commands/mod.rs:
```diff
@@ -1009,6 +1032,229 @@ impl DaemonControl {
+        // Loop will continue until either we find a suitable replacement or we have insufficient funds.
+        let mut rbf_vsize = 0_u64;
+        loop {
+            // Note that `rbf_vsize` strictly increases on each iteration as otherwise `min_fee`
+            // will ensure that the PSBT meets the required replacement fee and the loop will exit.
```
nit: i think the sentence is a bit off.

darosior commented on src/commands/mod.rs:
```diff
@@ -1009,6 +1032,229 @@ impl DaemonControl {
+            .max_sat_weight()
+            .try_into()
+            .expect("it must fit");
+        // Try with increasing feerate until fee paid by replacement transaction is high enough.
+        // Replacement fee must be at least:
```
With increasing *fee*. We do increase the feerate but what it's doing is increasing the fee, so i think it's better guidance for the reader to say that.

jp1ac4 replied:
Ah yes thanks, this was a leftover comment from the first approach increasing the feerate directly. Will change it.

jp1ac4 replied:
I've now updated this.

darosior commented on src/commands/mod.rs:
```diff
@@ -1009,6 +1032,229 @@ impl DaemonControl {
+        // Loop will continue until either we find a suitable replacement or we have insufficient funds.
+        let mut rbf_vsize = 0_u64;
+        loop {
+            // Note that `rbf_vsize` strictly increases on each iteration as otherwise `min_fee`
+            // will ensure that the PSBT meets the required replacement fee and the loop will exit.
+            let min_fee = mempool_entry.fees.descendant.to_sat() + rbf_vsize;
```
Actually `rbf_vsize` does not strictly increase, as you re-set it to the same value on every iteration below. We should have a test which exercises the case where there is >1 iteration, this would have caught the infinite loop here.

Suggestion for a fix:
```diff
diff --git a/src/commands/mod.rs b/src/commands/mod.rs
index c662300..d6ee1d9 100644
--- a/src/commands/mod.rs
+++ b/src/commands/mod.rs
@@ -1021,11 +1021,12 @@ impl DaemonControl {
         // Replacement fee must be at least:
         // sum of fees paid by original transactions + incremental feerate * replacement size.
         // Loop will continue until either we find a suitable replacement or we have insufficient funds.
-        let mut rbf_vsize = 0_u64;
-        loop {
+        let mut replacement_vsize = 0;
+        for incremental_feerate in 0.. {
             // Note that `rbf_vsize` strictly increases on each iteration as otherwise `min_fee`
             // will ensure that the PSBT meets the required replacement fee and the loop will exit.
-            let min_fee = mempool_entry.fees.descendant.to_sat() + rbf_vsize;
+            let min_fee =
+                mempool_entry.fees.descendant.to_sat() + replacement_vsize * incremental_feerate;
             let rbf_psbt = match self.create_spend_internal(
                 &destinations,
                 &candidate_coins,
@@ -1050,17 +1051,20 @@ impl DaemonControl {
                     return Err(e);
                 }
             };
-            rbf_vsize = unsigned_tx_max_vbytes(&rbf_psbt.psbt.unsigned_tx, max_sat_weight);
+            replacement_vsize = unsigned_tx_max_vbytes(&rbf_psbt.psbt.unsigned_tx, max_sat_weight);
 
+            // Make sure it satisfies RBF rule 4.
             if rbf_psbt
                 .psbt
                 .fee()
                 .expect("has already been sanity checked")
-                >= mempool_entry.fees.descendant + bitcoin::Amount::from_sat(rbf_vsize)
+                >= mempool_entry.fees.descendant + bitcoin::Amount::from_sat(replacement_vsize)
             {
                 return Ok(rbf_psbt);
             }
         }
+
+        unreachable!("We keep increasing the min fee until we run out of funds or satisfy rule 4.")
     }
 
     /// Trigger a rescan of the block chain for transactions involving our main descriptor between
```

jp1ac4 replied:
Do you mean there's an infinite loop in the case that `rbf_vsize` stays the same or always? In some of my tests, there was more than one iteration and, printing the value of `rbf_vsize`, I saw that it increased from one iteration to the next.

I might have overlooked something, but my thinking was that
```
rbf_fee_{i+1} >= minfee_{i+1} = descendant_fee + rbf_vsize_{i}
```
so that if `rbf_vsize_{i} >= rbf_vsize_{i+1}` then 
```
rbf_fee_{i+1} >= descendant_fee + rbf_vsize_{i+1}
```
and so we would exit the loop. So we only continue in the loop if `rbf_vsize` strictly increases.

darosior replied:
Yes, i'm not quite convinced `rbf_size` would always increase on every iteration.

jp1ac4 replied:
My reasoning above uses the `min_fee` parameter in `create_spend_internal` to ensure the replacement pays at least this fee. and should apply in any case that `rbf_vsize_{i} >= rbf_vsize_{i+1}`. As long as the `min_fee` parameter works as expected, I don't think there can be an infinite loop here.

But I may have overlooked something, so will make the change you suggested to be on the safe side.

jp1ac4 replied:
I've now updated this using your suggestion.

darosior commented on src/commands/mod.rs:
```diff
@@ -462,240 +447,63 @@ impl DaemonControl {
+        // Check the destination addresses are valid for the network and
         // sanity check each output's value.
-        let mut psbt_outs = Vec::with_capacity(destinations.len());
+        let mut destinations_checked: HashMap<bitcoin::Address, bitcoin::Amount> =
```
No need for the type declaration here

jp1ac4 replied:
I've removed this and a similar one in `rbf_psbt()`.

darosior reviewed (commented):
Haven't looked at the tests yet. This is looking good, i like the create_spend_internal interface, definitely paving the way for completely abstracting out the spend creation. Nice commits organization too, although i'd have made a ~~simple~~ single commit of 5fe6ae90d4191f3ce737d4e9532b956207d417d6 and 41d6a4bedf5269d7b1b280b069f9ddd31ecba40a so it can be reviewed using `--color-moved=dimmed-zebra`.

jp1ac4 commented on src/bitcoin/d/mod.rs:
```diff
@@ -944,7 +944,10 @@ impl BitcoinD {
-                    if confs == 0 && !conflicts.is_empty() && !self.is_in_mempool(&spending_txid) {
+                    if confs == 0
+                        && !conflicts.is_empty()
+                        && self.mempool_entry(&spending_txid).is_none()
```
I'll revert this change and follow your suggestion instead. It's in its own commit so should be straightforward.

jp1ac4 replied:
I've now reverted the change and redefined `is_in_mempool`.

jp1ac4 commented:
> i'd have made a simple commit of [5fe6ae9](https://github.com/wizardsardine/liana/commit/5fe6ae90d4191f3ce737d4e9532b956207d417d6) and [41d6a4b](https://github.com/wizardsardine/liana/commit/41d6a4bedf5269d7b1b280b069f9ddd31ecba40a) so it can be reviewed using `--color-moved=dimmed-zebra`.

Sure, I'll squash those two commits. I kept them separate just to keep options open as I wasn't sure what would be best.

jp1ac4 commented on src/commands/mod.rs:
```diff
@@ -1009,6 +1032,229 @@ impl DaemonControl {
+                *txid,
+            )));
+        }
+        let mempool_entry = self
+            .bitcoin
+            .mempool_entry(txid)
+            .ok_or(CommandError::RbfError(RbfErrorInfo::TxNotYetBroadcast(
+                *txid,
+            )))?;
```
That sounds good. Will change it to handle if tx is not in the mempool.

darosior replied:
I don't think there is a way to distinguish both.

Good point. I had overlooked that. You can use `gettxspendingprevout` to list the transactions spending a given set of coins. This can be a more generalistic way of getting the target min fee: always use `gettxspendingprevout` and query `getmempoolentry` for each of them. If the list is empty, assume it was purged and use a minfee of 0?

jp1ac4 replied:
Would it be possible to look up the `spend_txid` for each coin in the DB instead of using `gettxspendingprevout`?

jp1ac4 replied:
I've updated it now to get spending transactions from the DB.

Even if it's not in the DB or mempool, I always include the `txid` of the transaction to be replaced in the set of directly conflicting transactions so that the replacement's feerate will be higher. If the transaction to be replaced is not in the mempool, I make the assumption that the PSBT has been finalized so that we can get its fee and vsize, but perhaps I need to check this is the case as it's now possible to call the `rbfpsbt` command on an unsigned PSBT. I've added comments regarding this in the code.

darosior commented on src/commands/mod.rs:
```diff
@@ -780,6 +840,229 @@ impl DaemonControl {
+        // If cancel, the replacement will not have any destinations, only a change output.
+        let destinations: HashMap<bitcoin::Address, bitcoin::Amount> = if !is_cancel {
+            prev_derivs
+                .iter()
+                .filter_map(|(addr, amt, _)| {
+                    if prev_change_address != Some(addr) {
+                        Some((addr.clone(), *amt))
```
Probably a good idea to move `prev_change_address`'s `cloned()` from below to just above if you do that.

jp1ac4 replied:
I've now made this change.

darosior commented:
Just realized that actually this doesn't support RBF'ing a recovery transaction as in this case we need to set the right sequence in the inputs.

darosior commented on src/commands/mod.rs:
```diff
@@ -780,6 +840,229 @@ impl DaemonControl {
+                .map(|txin| txin.previous_output)
+                .collect();
+            let prev_coins = db_conn.coins_by_outpoints(prev_outpoints);
+            assert_eq!(prev_outpoints.len(), prev_coins.len());
```
Presumably the tx would have been dropped because its fees were below the mempool minfee so they would need to bump the fees before re-broadcasting it.

jp1ac4 replied:
I now check for outpoints for which we have no coin in the DB and return an error.

jp1ac4 commented on src/commands/mod.rs:
```diff
@@ -1009,6 +1032,229 @@ impl DaemonControl {
+                        return Err(CommandError::RbfError(RbfErrorInfo::TooLowFeerate(rate)));
+                    } else {
+                        min_feerate_vb
+                    }
```
This will now also error for cancel.

jp1ac4 commented on src/commands/mod.rs:
```diff
@@ -1009,6 +1032,229 @@ impl DaemonControl {
+            .spend_tx(txid)
+            .ok_or(CommandError::UnknownSpend(*txid))?;
+        let prev_coins = {
+            let prev_outpoints: &Vec<bitcoin::OutPoint> = &prev_psbt
```
I've now changed this.

jp1ac4 commented on doc/API.md:
```diff
@@ -257,6 +258,29 @@ This command does not return anything for now.
 
+### `rbfpsbt`
+
+Create PSBT to replace the given transaction using RBF.
```
I've added this.

jp1ac4 commented on src/commands/mod.rs:
```diff
@@ -1009,6 +1032,229 @@ impl DaemonControl {
+    /// value will be used as a change address and the others will be treated as non-change outputs.
+    ///
+    /// `is_cancel` indicates whether to "cancel" the transaction by removing non-change outputs from
+    /// the replacement or otherwise to keep the same non-change outputs and simply bump the fee.
```
I've updated the wording now so hopefully it's a bit clearer.

jp1ac4 commented on doc/API.md:
```diff
@@ -257,6 +258,29 @@ This command does not return anything for now.
+| Field    | Type   | Description                                                                              |
+| -------- | ------ | ---------------------------------------------------------------------------------------- |
+| `txid`   | string | Hex encoded txid of the Spend transaction to be replaced.                                |
+| `is_cancel`   | bool | Whether to "cancel" the transaction by removing non-change outputs from the replacement or otherwise to keep the same non-change outputs and simply bump the fee. If `true`, the only output of the RBF transaction will be change and the inputs will include at least one of the inputs from the previous transaction. If `false`, all inputs from the previous transaction will be  ...
```
I've updated this now.

jp1ac4 commented:
> Sure, I'll squash those two commits. I kept them separate just to keep options open as I wasn't sure what would be best.

I've now done this.

jp1ac4 commented:
I've now addressed the comments.

darosior commented:
@edouardparis can you have a quick look at 01d50c956ac0ad7ca7247fe7085c5ca8ab13ee0c?

jp1ac4 commented on src/commands/mod.rs:
```diff
@@ -171,6 +173,19 @@ pub enum InsaneFeeInfo {
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum RbfErrorInfo {
+    TooLowFeerate(u64),
```
I think it would be helpful to include the min feerate here as well as / instead of the chosen feerate.

jp1ac4 replied:
I haven't changed this yet. We could leave for a follow-up.

darosior commented on src/commands/mod.rs:
```diff
@@ -816,6 +831,265 @@ impl DaemonControl {
+            min_feerate_vb =
+                min_feerate_vb.max(conflict_feerate_vb.to_sat().checked_add(1).unwrap());
+        }
+        // Check replacement transaction's target feerate, if set, is high enough,
+        // and otherwise set it to the min feerate found above.
+        let feerate_vb = feerate_vb.unwrap_or(min_feerate_vb);
+        if feerate_vb < min_feerate_vb {
+            return Err(CommandError::RbfError(RbfErrorInfo::TooLowFeerate(
+                feerate_vb,
+            )));
+        }
```
Conversely, i don't think there is any reason for setting the feerate when using a cancel.

darosior commented on src/commands/mod.rs:
```diff
@@ -816,6 +831,265 @@ impl DaemonControl {
+        // Check replacement transaction's target feerate, if set, is high enough,
+        // and otherwise set it to the min feerate found above.
+        let feerate_vb = feerate_vb.unwrap_or(min_feerate_vb);
+        if feerate_vb < min_feerate_vb {
```
I was wrong here, it's already accounted for in `min_feerate_vb` through the `checked_add` above.

darosior commented:
I've addressed my comments in separate commits for you to review and squash them in the appropriate ones.
- 575025da04dc756ce8a154d26eb6ddae4e677df8, cf10f1396f45b2190f844916196a5faabc98ea8a and af4730b28da123887b83eebdbbfd1c0bb95f342d address https://github.com/wizardsardine/liana/pull/816#discussion_r1416863484 and https://github.com/wizardsardine/liana/pull/816#discussion_r1416882081
- de2f5197573504d27d5b91c006a95a38f0176593 addresses https://github.com/wizardsardine/liana/pull/816#discussion_r1416075264
- c663bb1f3c40e9ae9540cbf7d7ae0639d356d590 addresses https://github.com/wizardsardine/liana/pull/816#discussion_r1416097845
- 96f2a5dd002127d8d24fa2336fb93f69329af298 addresses https://github.com/wizardsardine/liana/pull/816#discussion_r1416909534

I've also played with this a bit more. I think this is good to go if you agree with my changes.

jp1ac4 commented on src/commands/mod.rs:
```diff
@@ -858,29 +868,32 @@ impl DaemonControl {
-            prev_coins
-        };
+        if !prev_psbt.unsigned_tx.is_explicitly_rbf() {
+            return Err(CommandError::RbfError(RbfErrorInfo::NotSignaling));
```
Should we rather check all the spending transactions in the mempool? Maybe using the `bip125-replaceable` field in `getmempoolentry`?

jp1ac4 replied:
I think the directly conflicting transactions should correspond to the `mempool_spenders` for the previous outpoints, but happy to come back to that in a follow-up.

darosior commented:
ACK 0ab00cd6580ddc1964f4f328a90ab9c5f3f87efa
